### PR TITLE
Fixed compilation errors for template deductions

### PR DESCRIPTION
**Pull request by doe300**
When trying to compile the current version with the new message-system, I get compilation errors failing to deduce a template argument.

TLDR; The compiler (`gcc version 6.4.1 20170727 (Red Hat 6.4.1-1) (GCC)`) fails to deduce an `unsigned int` template parameter from a `long unsigned int` value.

This PR fixes the error.

Full errors:

    /tmp/vc4asm/src/Parser.cpp: In member function ‘std::vector<exprValue> Parser::parseArgumentList(const string&, size_t)’:
    /tmp/vc4asm/src/Parser.cpp:1497:69: error: no matching function for call to ‘Parser::Fail(const msgTemplate<const char*, unsigned int, unsigned int>&, const char*, size_t&, std::vector<exprValue>::size_type)’
     Fail(MSG.FUNCTION_TOO_FEW_ARGS, name.c_str(), count, args.size());
                                                                     ^
    In file included from /tmp/vc4asm/src/Parser.h:12:0,
                 from /tmp/vc4asm/src/Parser.cpp:10:
    /tmp/vc4asm/src/AssembleInst.h:219:20: note: candidate: template<class ... A> void AssembleInst::Fail(msgTemplate<A ...>, A ...) const
    [[noreturn]] void Fail(const msgTemplate<A...> msg, A... a) const
                    ^~~~
    /tmp/vc4asm/src/AssembleInst.h:219:20: note:   template argument deduction/substitution failed:
    /tmp/vc4asm/src/Parser.cpp:1497:69: note:   inconsistent parameter pack deduction with ‘unsigned int’ and ‘long unsigned int’
     Fail(MSG.FUNCTION_TOO_FEW_ARGS, name.c_str(), count, args.size());
                                                                     ^
    /tmp/vc4asm/src/Parser.cpp:1506:57: error: no matching function for call to ‘Parser::Fail(const msgTemplate<const char*, unsigned int>&, const char*, size_t&)’
     Fail(MSG.FUNCTION_TOO_MANY_ARGS, name.c_str(), count);
                                                         ^
    In file included from /tmp/vc4asm/src/Parser.h:12:0,
                 from /tmp/vc4asm/src/Parser.cpp:10:
    /tmp/vc4asm/src/AssembleInst.h:219:20: note: candidate: template<class ... A> void AssembleInst::Fail(msgTemplate<A ...>, A ...) const
    [[noreturn]] void Fail(const msgTemplate<A...> msg, A... a) const
                    ^~~~
    /tmp/vc4asm/src/AssembleInst.h:219:20: note:   template argument deduction/substitution failed:
    /tmp/vc4asm/src/Parser.cpp:1506:57: note:   inconsistent parameter pack deduction with ‘unsigned int’ and ‘long unsigned int’
     Fail(MSG.FUNCTION_TOO_MANY_ARGS, name.c_str(), count);
                                                         ^
    /tmp/vc4asm/src/Parser.cpp: In member function ‘void Parser::doMACRO(std::unordered_map<std::__cxx11::basic_string<char>, Parser::macro>::const_iterator)’:
    /tmp/vc4asm/src/Parser.cpp:1601:69: error: no matching function for call to ‘Parser::Fail(const msgTemplate<const char*, unsigned int>&, const char*, std::vector<std::__cxx11::basic_string<char> >::size_type)’
      Fail(MSG.MACRO_TOO_MANY_ARGS, m->first.c_str(), argnames.size());
                                                                     ^
    In file included from /tmp/vc4asm/src/Parser.h:12:0,
                 from /tmp/vc4asm/src/Parser.cpp:10:
    /tmp/vc4asm/src/AssembleInst.h:219:20: note: candidate: template<class ... A> void AssembleInst::Fail(msgTemplate<A ...>, A ...) const
    [[noreturn]] void Fail(const msgTemplate<A...> msg, A... a) const
                    ^~~~
    /tmp/vc4asm/src/AssembleInst.h:219:20: note:   template argument deduction/substitution failed:
    /tmp/vc4asm/src/Parser.cpp:1601:69: note:   inconsistent parameter pack deduction with ‘unsigned int’ and ‘long unsigned int’
      Fail(MSG.MACRO_TOO_MANY_ARGS, m->first.c_str(), argnames.size());
                                                                     ^
    /tmp/vc4asm/src/Parser.cpp:1605:81: error: no matching function for call to ‘Parser::Fail(const msgTemplate<const char*, unsigned int, unsigned int>&, const char*, std::vector<std::__cxx11::basic_string<char> >::size_type, std::vector<exprValue>::size_type)’
      Fail(MSG.MACRO_TOO_FEW_ARGS, m->first.c_str(), argnames.size(), args.size());
                                                                                 ^
    In file included from /tmp/vc4asm/src/Parser.h:12:0,
                 from /tmp/vc4asm/src/Parser.cpp:10:
    /tmp/vc4asm/src/AssembleInst.h:219:20: note: candidate: template<class ... A> void AssembleInst::Fail(msgTemplate<A ...>, A ...) const
    [[noreturn]] void Fail(const msgTemplate<A...> msg, A... a) const
                    ^~~~
    /tmp/vc4asm/src/AssembleInst.h:219:20: note:   template argument deduction/substitution failed:
    /tmp/vc4asm/src/Parser.cpp:1605:81: note:   inconsistent parameter pack deduction with ‘unsigned int’ and ‘long unsigned int’
      Fail(MSG.MACRO_TOO_FEW_ARGS, m->first.c_str(), argnames.size(), args.size());

### CODE REVIEW

**Comment by maazl** [comment]
Well, I already have run into this pitfall at other locations where changing the message template was no reasonable option. It is obviously a restriction of the C++ standard, that no implicit type conversions take place except for base class conversion when the called function is _not_ a member function of the msgTemplate class.
As soon as find an appropriate work around without loosing type safety I will apply it.

BTW: now the printf format string are no longer consistent. They should be %zu for size_t. It makes no difference on my platforms (Mint17@i686 & Jessie@armv6l) for the same reason that I did not get compilation errors.

**Comment by doe300** [comment]
The error only occurs on 64-bit systems, since for 32-bit systems (like most of the Raspbery Pis), `size_t` is aliased to `unsigned int`, but for 64-bit systems it is an alias for `long unsigned int`.
